registry/remote/auth/example_test.go

```diff
@@ -0,0 +1,220 @@
+				w.Header().Set("Www-Authenticate", challenge)
+				w.WriteHeader(http.StatusUnauthorized)
+			}
+		case "/clientConfig":
+			wantedAuthHeader := "Basic " + base64.StdEncoding.EncodeToString([]byte(username+":"+password))
+			authHeader := r.Header.Get("Authorization")
+			if authHeader != wantedAuthHeader {
+				w.Header().Set("Www-Authenticate", `Basic realm="Test Server"`)
+				w.WriteHeader(http.StatusUnauthorized)
+			}
+		case "/simple":
+			w.WriteHeader(http.StatusOK)
+		default:
+			w.WriteHeader(http.StatusNotAcceptable)
+		}
+	}))
+	defer ts.Close()
+	host = ts.URL
+	simpleDir = fmt.Sprintf("%s/simple", host)
+	basicAuthDir = fmt.Sprintf("%s/basicAuth", host)
+	accessTokenDir = fmt.Sprintf("%s/accessToken", host)
+	clientConfigDir = fmt.Sprintf("%s/clientConfig", host)
+	http.DefaultClient = ts.Client()
+
+	os.Exit(m.Run())
+}
+
+// ExampleClient_Do_minimalClient gives an example of a minimal working client.
+func ExampleClient_Do_minimalClient() {
+	var client auth.Client
+	req, err := http.NewRequest(http.MethodGet, simpleDir, nil)
+	if err != nil {
+		panic(err)
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		panic(err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		panic(resp.StatusCode)
+	}
+
+	fmt.Println(resp.StatusCode)
+	// Output:
+	// 200
+}
+
+// ExampleClient_Do_basicAuth gives an example of using client with credentials.
+func ExampleClient_Do_basicAuth() {
+	client := &auth.Client{
+		Credential: func(ctx context.Context, reg string) (auth.Credential, error) {
+			return auth.Credential{
+				Username: username,
+				Password: password,
+			}, nil
+		},
+	}
+	req, err := http.NewRequest(http.MethodGet, basicAuthDir, nil)
+	if err != nil {
+		panic(err)
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		panic(err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		panic(resp.StatusCode)
+	}
+
+	fmt.Println(resp.StatusCode)
+	// Output:
+	// 200
+}
+
+// ExampleClient_Do_withAccessToken gives an example of using client with an access token.
+func ExampleClient_Do_withAccessToken() {
+	client := &auth.Client{
+		Credential: func(ctx context.Context, reg string) (auth.Credential, error) {
+			return auth.Credential{
+				AccessToken: accessToken,
+			}, nil
+		},
+	}
+	req, err := http.NewRequest(http.MethodGet, accessTokenDir, nil)
+	if err != nil {
+		panic(err)
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		panic(err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		panic(resp.StatusCode)
+	}
+
+	fmt.Println(resp.StatusCode)
+	// Output:
+	// 200
+}
+
+// ExampleClient_Do_clientConfiguration shows the client configurations available,
+// including using cache, setting user agent and configuring OAuth2.
+func ExampleClient_Do_clientConfiguration() {
+	client := &auth.Client{
+		Credential: func(ctx context.Context, reg string) (auth.Credential, error) {
+			return auth.Credential{
+				Username: username,
+				Password: password,
+			}, nil
+		},
```

Comment:
@shizhMSFT  @Wwwsylvia Should we use a factory function to construct the `client.Credential` here? It would improve the user experience.

```go
// added to auth/credential.go
func FromStatic(cred Credential) func(ctx context.Context, reg string) (Credential, error) {
	return func(ctx context.Context, reg string) (Credential, error) {
		return cred, nil
	}
}
```

So code here can be simplified as below:
```
Credential: auth.FromStatic(auth.Credential{
	Username: username,
	Password: password,
}),
```

Reply:
@qweeah I think that would be useful, and we can do the registry validation for users.